src/coprocessor/dag/batch_executor/interface.rs

```diff
@@ -0,0 +1,166 @@
+    pub fn with_default_config(columns_info: Vec<ColumnInfo>) -> Self {
+        Self::new(columns_info, EvalConfig::default())
+    }
+}
+
+impl Deref for BatchExecutorContext {
+    type Target = BatchExecutorContextInner;
+
+    fn deref(&self) -> &BatchExecutorContextInner {
+        &self.0
+    }
+}
+
+impl crate::util::AssertSend for BatchExecutorContext {}
+
+impl crate::util::AssertSync for BatchExecutorContext {}
+
+/// The actual inner structure for `BatchExecutorContext`. Normally this structure should be
+/// accessed via `BatchExecutorContext` and should not be constructed directly.
+pub struct BatchExecutorContextInner {
+    /// The column info of base physical table schemas.
+    // TODO: Actually we may need different columns_info for different executors. For example,
+    // the executor above the aggregation executor's (e.g. Projection Executor) schema should be
+    // the aggregation executor's schema instead of base schema.
+    pub columns_info: Vec<ColumnInfo>,
+
+    // TODO: This is really a execution config, alhtough called eval config.
+    pub config: EvalConfig,
+}
+
+/// Data to be flowed between parent and child executors' single `next_batch()` invocation.
+///
+/// Note: there are other data flow between executors, like metrics and output statistics.
+/// However they are flowed at once, just before response, instead of each step during execution.
+/// Hence they are not covered by this structure. See `BatchExecuteMetaData`.
+///
+/// It is only `Send` but not `Sync` because executor returns its own data copy. However `Send`
+/// enables executors to live in different threads.
+///
+/// It is designed to be used in new generation executors, i.e. executors support batch execution.
+/// The old executors will not be refined to return this kind of result.
+pub struct BatchExecuteResult {
+    /// The columns data generated during this invocation. Note that empty column data doesn't mean
+    /// that there is no more data. See `is_drained`.
+    pub data: LazyBatchColumnVec,
+
+    /// The warnings generated during this invocation.
+    // TODO: It can be more general, e.g. `ExecuteWarnings` instead of `EvalWarnings`.
+    // TODO: Should be recorded by row.
+    pub warnings: EvalWarnings,
+
+    /// Whether or not there is no more data.
+    ///
+    /// This structure is a `Result`. When it is:
+    /// - `Ok(false)`: The normal case, means that there could be more data. The caller should
+    ///                continue calling `next_batch()` although for each call the returned data may
+    ///                be empty.
+    /// - `Ok(true)`:  Means that the executor is drained and no more data will be returned in
+    ///                future. However there could be some (last) data in the `data` field this
+    ///                time. The caller should NOT call `next_batch()` any more.
+    /// - `Err(_)`:    Means that there is an error when trying to retrieve more data. In this case,
```

Comment:
Could we return `Result<BatchExecuteResult>` for the function `next_batch` instead? so the `is_drained` could just be defined as a boolean value, and we needn't add so much comments to introduce it.

Reply:
No, because when there is an error, we still want results. Result<xxx> makes it an `either` instead of `both`.

Reply:
One error is enough? Or we need an array like `warnings`? And in this case, I prefer to define a new attribute to store it instead of put it together with `is_drained`

Reply:
@AndreMouche When there is an error, we *may* need it (but not 100%). For example, we may scan 100 rows first and meet an error, but later if we only need 50 rows from these 100 rows then the error should be ignored.